src/input/datetime.rs

```diff
@@ -69,6 +69,61 @@ impl<'a> From<&'a PyTime> for EitherTime<'a> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum EitherTimedelta<'a> {
+    Raw(Duration),
+    Py(&'a PyDelta),
+}
+
+impl<'a> From<Duration> for EitherTimedelta<'a> {
+    fn from(timedelta: Duration) -> Self {
+        Self::Raw(timedelta)
+    }
+}
+
+impl<'a> From<&'a PyDelta> for EitherTimedelta<'a> {
+    fn from(timedelta: &'a PyDelta) -> Self {
+        Self::Py(timedelta)
+    }
+}
+
+impl<'a> EitherTimedelta<'a> {
+    pub fn as_raw(&self) -> PyResult<Duration> {
+        match self {
+            Self::Raw(timedelta) => Ok(timedelta.clone()),
+            Self::Py(py_timedelta) => Ok(Duration::new(
+                true,
+                py_timedelta
+                    .getattr(pyo3::intern!(py_timedelta.py(), "days"))?
+                    .extract()?,
+                py_timedelta
+                    .getattr(pyo3::intern!(py_timedelta.py(), "seconds"))?
+                    .extract()?,
+                py_timedelta
+                    .getattr(pyo3::intern!(py_timedelta.py(), "microseconds"))?
+                    .extract()?,
+            )),
+        }
+    }
+
+    pub fn try_into_py(self, py: Python<'_>) -> PyResult<PyObject> {
+        let timedelta = match self {
+            Self::Py(timedelta) => Ok(timedelta),
+            Self::Raw(duration) => {
+                let sign = if duration.positive { 1 } else { -1 };
+                PyDelta::new(
+                    py,
+                    sign * duration.day as i32,
+                    sign * duration.second as i32,
+                    sign * duration.microsecond as i32,
```

Comment:
I think we should raise an error if any of these values exceed `i32::MAX` - I think currently the value wraps around but the timedelta doesn't mean much at that point.